src/calendar/model/CalendarModel.ts

```diff
@@ -265,89 +277,227 @@ export class CalendarModel {
+			const updatedProgenitor = clone(dbTarget.progenitor)
+			updatedProgenitor.repeatRule = repeatRuleWithExcludedAlteredInstances(updatedProgenitor, [updateEvent.recurrenceId], this.zone)
+			await this.doUpdateEvent(dbTarget.progenitor, updatedProgenitor)
+		} else if (updateEvent.recurrenceId == null && updateEvent.repeatRule != null && dbTarget.alteredInstances.length > 0) {
+			// request to add the progenitor to the calendar. we have to exclude all altered instances that are known to us from it.
+			updateEvent.repeatRule = repeatRuleWithExcludedAlteredInstances(
+				updateEvent,
+				dbTarget.alteredInstances.map((r) => r.recurrenceId),
+				this.zone,
+			)
+		}
+		let calendarGroupRoot
+		try {
+			calendarGroupRoot = await this.entityClient.load(CalendarGroupRootTypeRef, dbTarget.ownerGroup)
+		} catch (e) {
+			if (!(e instanceof NotFoundError) && !(e instanceof NotAuthorizedError)) throw e
+			console.log(TAG, "tried to create new progenitor or got new altered instance for progenitor in nonexistent/inaccessible calendar, ignoring")
+			return
+		}
+		return await this.doCreate(updateEvent, "", calendarGroupRoot, alarms)
+	}
 
-			if (filterInt(dbEvent.sequence) < filterInt(event.sequence)) {
-				await this.updateEventWithExternal(dbEvent, event).then(noOp)
-				return
-			}
-		} else if (calendarData.method === CalendarMethod.CANCEL) {
-			if (dbEvent.organizer == null || dbEvent.organizer.address !== sender) {
-				console.log(TAG, "CANCEL sent not by organizer, ignoring")
-				return
-			}
-			await this.entityClient.erase(dbEvent)
+	/** Someone replied whether they attend an event or not. this MUST be applied to all instances in our
+	 * model since we keep attendee lists in sync for now. */
+	private async processCalendarReply(sender: string, dbEvent: CalendarEvent, updateEvent: CalendarEvent): Promise<void> {
+		console.log("processing calendar reply")
+		// first check if the sender of the email is in the attendee list
+		const replyAttendee = findAttendeeInAddresses(updateEvent.attendees, [sender])
+
+		if (replyAttendee == null) {
+			console.log(TAG, "Sender is not among attendees, ignoring", replyAttendee)
+			return
+		}
+
+		const newEvent = clone(dbEvent)
+		// check if the attendee is still in the attendee list of the latest event
+		const dbAttendee = findAttendeeInAddresses(newEvent.attendees, [replyAttendee.address.address])
+
+		if (dbAttendee == null) {
+			console.log(TAG, "attendee was not found", dbEvent._id, replyAttendee)
 			return
 		}
+
+		dbAttendee.status = replyAttendee.status
+		await this.doUpdateEvent(dbEvent, newEvent)
+	}
+
+	/** handle an event cancellation - either the whole series (progenitor got cancelled)
+	 * or the altered occurrence. */
+	private async processCalendarCancellation(dbEvent: CalendarEvent): Promise<void> {
+		console.log("processing cancellation")
+		// not having UID is technically an error, but we'll do our best (the event came from the server after all)
```

Comment:
we can't not have uid here because we've found this event by uid

Reply:
yes, but the compiler does not prevent us from passing an event without UID here. There also might be a bug with the index that allows setting the uid to null after the event was added to the index.

we could just throw here in that case, but I think deleting the event even though it's technically invalid would be enough to not degrade functionality too much.